### PR TITLE
support InfiniteTimeSpan in Select

### LINKED ISSUE #72458: New Socket.Select TimeSpan-based overload is broken for Timeout.InfiniteTimeSpan
The existing overload of `Socket.Select` accepts an `int` that represents the number of microseconds.  Like most int-based timeouts, it treats -1 as meaning infinite.  But Timeout.InfiniteTimeSpan is -1 milliseconds, not -1 microseconds.  As such, when the new `Socket.Select` overload that takes a TimeSpan tries to convert Timeout.InfiniteTimeSpan to microseconds, it throws with:
```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'timeout')
   at System.Net.Sockets.Socket.ToTimeoutMicroseconds(TimeSpan timeout)
   at System.Net.Sockets.Socket.Select(IList checkRead, IList checkWrite, IList checkError, TimeSpan timeout)
```

### PR DESCRIPTION
**Pull request by wfurt**
This change adds special handling for `Timeout.InfiniteTimeSpan` so it can be passed to Select without throwing. 
fixes #72458

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 1]
```diff
@@ -2228,6 +2228,11 @@ public static void Select(IList? checkRead, IList? checkWrite, IList? checkError
 
         private static int ToTimeoutMicroseconds(TimeSpan timeout)
         {
+            if (timeout == Timeout.InfiniteTimeSpan)
+            {
+                return -1;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2234]: While this helps to make TimeSpan.InfiniteTimeSpan / TimeSpan.FromMilliseconds(-1) be treated as infinite, it feels weird to me that TimeSpan.FromMicroseconds(-1) is still also going to be treated as infinite.  We also need to decide if we care about values that are truncated to the same value, e.g. TimeSpan.FromMilliseconds(-1.1) is also going to be treated as infinite by other APIs that accept TimeSpans.

(And regardless of what we do here, the XML comments on the above method need to be updated, as they talk about -1 microseconds.)

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2234]: There is really no practical use for any negative numbers besides infinite, right? We can document that -1 (from Microseconds) and InfiniteTimeSpan are both treated equally. And we can reject rest of negative numbers IMHO.

If you have other APIs in mind for consistency please let me know.

Comment (ManickaP) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2234]: IMO, we should accept only `Timeout.InfiniteTimeSpan` from the negative numbers and converted in whatever value represents infinite in the underlying function. 

Also this method is called from `Poll`, so we should make sure the behavior for infinite is the same.

Finally, the xml comment of `Select` above should drop the detail about -1 microsecond as Stephen mentioned.

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2234]: updated + new test

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 2]
```diff
@@ -2217,19 +2217,29 @@ public static void Select(IList? checkRead, IList? checkWrite, IList? checkError
         /// <param name="checkRead">An <see cref="IList"/> of <see cref="Socket"/> instances to check for readability.</param>
         /// <param name="checkWrite">An <see cref="IList"/> of <see cref="Socket"/> instances to check for writability.</param>
         /// <param name="checkError">An <see cref="IList"/> of <see cref="Socket"/> instances to check for errors.</param>
-        /// <param name="timeout">The timeout value. A value equal to -1 microseconds indicates an infinite timeout.</param>
+        /// <param name="timeout">The timeout value. A value equal to <see cref="Timeout.InfiniteTimeSpan"/> microseconds indicates an infinite timeout.</param>
```
Comment (ManickaP) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2220]:
```suggestion
        /// <param name="timeout">The timeout value. A value equal to <see cref="Timeout.InfiniteTimeSpan"/> indicates an infinite timeout.</param>
```

**Review by ManickaP — APPROVED** [review]
Thanks.

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 3]
```diff
@@ -2167,7 +2167,7 @@ public bool Poll(int microSeconds, SelectMode mode)
         /// does not block and the connection has failed, or if OutOfBandInline is not set and out-of-band data is available.
         /// Otherwise, it returns <see langword="false"/>.
         /// </returns>
-        /// <exception cref="ArgumentOutOfRangeException"><paramref name="timeout"/> is less than -1 milliseconds or greater than <see cref="int.MaxValue"/> milliseconds.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="timeout"/> was negative or greater thanTimeSpan.FromMicroseconds(int.MaxValue).</exception>
```
Comment (ManickaP) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:2170]:
```suggestion
        /// <exception cref="ArgumentOutOfRangeException"><paramref name="timeout"/> was negative or greater than TimeSpan.FromMicroseconds(int.MaxValue).</exception>
```

**Pull request merged**